monitoring/opencensus/metrics.go

```diff
@@ -0,0 +1,327 @@
+func (ocmf MetricFactory) NewCounter(name, help string, labelNames ...string) monitoring.Counter {
+	//TODO(dazwilkin) What View Aggregation is best for "Counter"? (sum?)
+	glog.Infof("[Counter] %s", name)
+	measure := createMeasureAndView(ocmf.Prefix, name, help, view.Sum(), labelNames)
+	return &Counter{
+		labelNames: labelNames,
+		measure:    measure,
+	}
+}
+
+// NewGauge creates a new Gauge object backed by OpenCensus
+func (ocmf MetricFactory) NewGauge(name, help string, labelNames ...string) monitoring.Gauge {
+	//TODO(dazwilkin) What View Aggregation is best for "Gauge"? (count+sum? lastvalue?)
+	glog.Infof("[Gauge] %s", name)
+	measure := createMeasureAndView(ocmf.Prefix, name, help, view.Sum(), labelNames)
+	return &Gauge{
+		labelNames: labelNames,
+		measure:    measure,
+	}
+}
+
+// Ref: https://github.com/google/trillian/blob/master/monitoring/prometheus/metrics.go
+// buckets returns a reasonable range of histogram upper limits for most
+// latency-in-seconds usecases.
+func buckets() []float64 {
+	// These parameters give an exponential range from 0.04 seconds to ~1 day.
+	num := 300
+	b := 1.05
+	scale := 0.04
+
+	r := make([]float64, 0, num)
+	for i := 0; i < num; i++ {
+		r = append(r, math.Pow(b, float64(i))*scale)
+	}
+	return r
+}
+
+// NewHistogram creates a new Histogram object backed by Prometheus.
+func (ocmf MetricFactory) NewHistogram(name, help string, labelNames ...string) monitoring.Histogram {
+	//TODO(dazwilkin) How is an OpenCensus Distribution treated by Stackdriver?
+	glog.Infof("[Histogram] %s", name)
+	measure := createMeasureAndView(ocmf.Prefix, name, help, view.Distribution(buckets()...), labelNames)
+	return &Histogram{
+		labelNames: labelNames,
+		measure:    measure,
+	}
+}
+
+// Counter is a wrapper around OpenCensus object
+type Counter struct {
+	labelNames []string
+	measure    *stats.Float64Measure
+}
+
+// Inc adds 1 to a counter.
+func (c *Counter) Inc(labelVals ...string) {
+	c.Add(1.0, labelVals...)
+}
+
+// Add adds the given amount to a counter.
+func (c *Counter) Add(val float64, labelVals ...string) {
+	//TODO(dazwilkin) Are negative values permitted? Think not.
+	// Nothing to do
+	if val <= 0.0 {
+		return
+	}
+	if err := forAllLabelsAValue(c.labelNames, labelVals); err != nil {
+		glog.Error(err.Error())
+		return
+	}
+	ctx := context.TODO()
+	ctx = assignValuesToLabels(ctx, c.labelNames, labelVals)
+	stats.Record(ctx, c.measure.M(val))
+}
+
+// Value returns the amount of a counter.
+func (c *Counter) Value(labelVals ...string) float64 {
+	if err := forAllLabelsAValue(c.labelNames, labelVals); err != nil {
+		glog.Error(err.Error())
+		return 0.0
+	}
+	glog.Error("Unable to return values for counters")
+	return 0.0
+}
+
+// Gauge is a wrapper around an OpenCensus measurement and view
+type Gauge struct {
+	labelNames []string
+	measure    *stats.Float64Measure
+}
+
+// Inc adds 1 to the gauge
+func (g *Gauge) Inc(labelVals ...string) {
```

Comment:
I don't understand this comment but I think perhaps you allude to this issue:

One challenge implementing the Trillian metric interface with OpenCensus is that the interface assumes the possibility of tracking state of the measure. Unlike with Prometheus (where this is possible), OpenCensus is a proxy (for one or more monitoring services) and the proxy simply dispatches the current value to the configured backends without recording the current measure's value.

As a result, it's possible to increment Counters, and set the value of Gauges. But it's not possible to get the value of either, and thus it's not possible to increment, decrement of add values to e.g. Gauges because this requires retaining knowledge of the current value.

If the interface is changed to reflect this, then the Prometheus tests that assume this, will break. As it is, the OpenCensus tests break because it's not possible to get the measure values.